core/deflator.js

```diff
@@ -0,0 +1,84 @@
+/*
+ * noVNC: HTML5 VNC client
+ * Copyright (C) 2020 The noVNC Authors
+ * Licensed under MPL 2.0 (see LICENSE.txt)
+ *
+ * See README.md for usage and integration instructions.
+ */
+
+import { deflateInit, deflate } from "../vendor/pako/lib/zlib/deflate.js";
+import { Z_FULL_FLUSH } from "../vendor/pako/lib/zlib/deflate.js";
+import ZStream from "../vendor/pako/lib/zlib/zstream.js";
+
+export default class Deflator {
+    constructor() {
+        this.strm = new ZStream();
+        this.chunkSize = 1024 * 10 * 10;
+        this.strm.output = new Uint8Array(this.chunkSize);
+        this.windowBits = 5;
+
+        deflateInit(this.strm, this.windowBits);
+    }
+
+    deflate(inData) {
+        this.strm.input = inData;
+        this.strm.avail_in = this.strm.input.length;
+        this.strm.next_in = 0;
+        this.strm.next_out = 0;
+        this.strm.avail_out = this.chunkSize;
+
+        let stream = this.strm;
+
+        // Use Z_FULL_FLUSH be more generic and server bug compatible
+        let lastRet = deflate(stream, Z_FULL_FLUSH);
+        let outData = new Uint8Array(stream.output.buffer, 0, stream.next_out);
+
+        /* If there is more data to read after our chunk we need to
+           continue reading until finished */
+        if (lastRet < 0) {
+            throw new Error("zlib deflate failed");
+        }
+
+        if (stream.avail_in > 0) {
+            // Read chunks until done
+
+            let chunks = [outData];
+            let totalLen = outData.length;
+            do {
+                stream.output = new Uint8Array(this.chunkSize);
```

Comment:
It seems I cannot comment on things GitHub considers "outdated", but this comment doesn't seem to be addressed:

> You need to reset output on the next call to deflate().